db/chained_delegating_reader.hh

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/shared_future.hh>
+
+#include "flat_mutation_reader.hh"
+
+// A reader which allows to insert a deferring operation before reading.
+// All calls will first wait for a future to resolve, then forward to a given underlying reader.
+class chained_delegating_reader : public flat_mutation_reader::impl {
+    flat_mutation_reader _underlying;
+    shared_future<> _ready;
+    std::function<void()> _on_destroyed;
+    
+public:
+    chained_delegating_reader(future<> ready, reader_permit permit, flat_mutation_reader underlying, std::function<void()> on_destroyed = []{})
+        : impl(underlying.schema(), std::move(permit))
+        , _underlying(std::move(underlying))
+        , _ready(std::move(ready))
+        , _on_destroyed(std::move(on_destroyed))
+    { }
+
+    ~chained_delegating_reader() {
+        _on_destroyed();
+    }
+
+    virtual future<> fill_buffer(db::timeout_clock::time_point timeout) override {
+        if (!_ready.available()) {
+            return _ready.get_future(timeout).then([this, timeout] {
+                return fill_buffer(timeout);
+            });
+        }
+        if (is_buffer_full()) {
+            return make_ready_future<>();
+        }
+        return to_reference(_underlying).fill_buffer(timeout).then([this] {
+            _end_of_stream = to_reference(_underlying).is_end_of_stream();
+            to_reference(_underlying).move_buffer_content_to(*this);
+        });
+    }
+
+    virtual future<> fast_forward_to(position_range pr, db::timeout_clock::time_point timeout) override {
+        if (!_ready.available()) {
+            return _ready.get_future(timeout).then([this, timeout, pr = std::move(pr)] () mutable {
+                return fast_forward_to(pr, timeout);
+            });
+        }
+        _end_of_stream = false;
+        forward_buffer_to(pr.start());
+        return to_reference(_underlying).fast_forward_to(std::move(pr), timeout);
+    }
+
+    virtual future<> next_partition() override {
+        if (!_ready.available()) {
+            return make_ready_future<>();
```

Comment:
I'm unsure whether this is correct (returning immediately). It looks it was done deliberately in the original code, so we left it this way.

Reply:
First, let's observe that when you call next_partition() on a new stream, it is always no-op. That's because until the first partition start is consume, the client side of the reader is not yet inside the partition, and next_partition() is not supposed to skip over it, it is supposed to arrive at it. Now, if `!_ready.available()` then the reader is not yet constructed, but we know that when it will, there will be nothing to do. So we can pretend that there is nothing to do already.